posthog/hogql_queries/insights/retention_query_runner.py

```diff
@@ -0,0 +1,380 @@
+            sample_expr = ast.SampleExpr(sample_value=ast.RatioExpr(left=ast.Constant(value=sampling_factor)))
+            result.select_from.sample = sample_expr
+
+        return result
+
+    def target_field(self) -> str:
+        return "e.person_id as target"
+
+    def get_timestamp_field(self) -> str:
+        start_of_interval_sql = self.get_start_of_interval_sql(
+            source=f"{self.EVENT_TABLE_ALIAS}.timestamp",
+        )
+        if self.event_query_type == RetentionQueryType.TARGET:
+            return f"DISTINCT {start_of_interval_sql} AS event_date"
+        elif self.event_query_type == RetentionQueryType.TARGET_FIRST_TIME:
+            return f"min({start_of_interval_sql}) as event_date"
+        else:
+            return f"{start_of_interval_sql} AS event_date"
+
+    def _get_entity_query(self, entity: Entity):
+        if entity.type == TREND_FILTER_TYPE_ACTIONS and entity.id is not None:
+            action = Action.objects.get(pk=entity.id)
+            return action_to_expr(
+                action,
+            )
+        elif entity.type == TREND_FILTER_TYPE_EVENTS:
+            if entity.id is None:
+                return ast.Constant(value=True)
+            else:
+                return ast.CompareOperation(
+                    op=ast.CompareOperationOp.Eq,
+                    left=ast.Field(chain=[self.EVENT_TABLE_ALIAS, "event"]),
+                    right=ast.Constant(value=entity.id),
+                )
+        else:
+            return ast.CompareOperation(
+                op=ast.CompareOperationOp.Eq,
+                left=ast.Field(chain=[self.EVENT_TABLE_ALIAS, "event"]),
+                right=ast.Constant(value=PAGEVIEW_EVENT),
+            )
+
+    def _get_date_filter(self):
+        query = (
+            f"event_date >= {{{self.event_query_type}_start_date}} AND event_date <= {{{self.event_query_type}_end_date}}"
+            if self.event_query_type == RetentionQueryType.TARGET_FIRST_TIME
+            else f"{self.EVENT_TABLE_ALIAS}.timestamp >= {{{self.event_query_type}_start_date}} AND {self.EVENT_TABLE_ALIAS}.timestamp <= {{{self.event_query_type}_end_date}}"
+        )
+        start_date = self.query_date_range.date_from()
+        end_date = (
+            (self.query_date_range.date_from() + self.query.retentionFilter.period_increment)
+            if self.query.retentionFilter == TRENDS_LINEAR  # .display == linear TODO: ???
+            and self.event_query_type == RetentionQueryType.TARGET
+            else self.query_date_range.date_to()
+        )
+        if self.query.retentionFilter.period != "Hour":
+            start_date = start_date.replace(hour=0, minute=0, second=0, microsecond=0)
+            end_date = end_date.replace(hour=0, minute=0, second=0, microsecond=0)
+        params = {
+            f"{self.event_query_type}_start_date": start_date,
+            f"{self.event_query_type}_end_date": end_date,
+        }
+        return query, params
```

Comment:
Can we use `QueryDateRange` here? It performs a lot of logic around setting the timestamps on dates dependent on the interval set etc, which probably covers a few more use cases than here